### PR TITLE
Module 2 Capstone Project

### PR DESCRIPTION
**Pull request by laiifuu**
## Hi there 👋 

For this project, @Zuhaib042 , @KareemWilson  and I (@laiifuu) worked to bring to you TV Zone, which is a website to check out information about your favorite shows. We built this project using the [TV Maze API](https://www.tvmaze.com/api) to fetch the information about the shows, and the [Involvement API](https://www.notion.so/Involvement-API-869e60b5ad104603aa6db59e08150270) to store the user's likes, comments and reservations.

Here's a summary of what we have accomplished:
- We built 3 interfaces: the home page, the the comments popup and the reservations popup following the wireframe provided. 
- We retrieved the data from both APIs as the page loads and displayed the shows and likes.
- We implemented the liking, commenting and reserving functionalities by sending data to the `involvement API` to be stored.
- We implemented the popup windows for both the reservations and comments sections.
- We implemented 3 counter functions for: 
    - The number of items (home).
    - The number of comments (comments popup).
    - The number of reservations (reservations popup).
- We added tests using `Jest` for each of the counter functions.

### CODE REVIEW

File: dist/index.html [hunk 1]
```diff
@@ -0,0 +1,63 @@
+<!DOCTYPE html>
+<html lang="en">
+
+<head>
+  <meta charset="UTF-8" />
+  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
+  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
+  <script src="index.js" type="module" defer></script>
+  <!-- <script src="https://cdn.tailwindcss.com"></script> -->
```
Comment (ichala) [inline dist/index.html:9]: - [ ] To keep your code clean, kindly get rid of any commented code or not helpful comments in your repo 🙏 
- [ ] you guys have an error in your website caused by calling index.js in your <head> tag in the index.html file , besides that the file src is wrong, since we are using webpack, it will be responsible for calling the index.js or (main.js as you guys decided to name it) inside the HTML file, so just remove the importing line `<script src="index.js" type="module" defer></script>` and the error will be fixed 👌 

image

File: src/utilities/homePageUtilities.js [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+  body,
+  url,
+) {
+  const previousPage = pageNumber - 1;
+  if (previousPage > 0) {
+    pageNum.innerHTML = previousPage;
+    showsDiv.innerHTML = '';
+    displayShows(shows, previousPage, showsHeader, showsDiv, url, body);
+  }
+}
```
Comment (ichala) [inline src/utilities/homePageUtilities.js:116]: - [ ] as part of the requirement, you guys should stick with ES6 as you did with most of your functions, kindly update `function loadPrevious` to be an arrow function 🙏

File: src/utilities/homePageUtilities.js [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+  body,
+  likesApiURL,
+) {
+  const nextPage = pageNumber + 1;
+  if (nextPage < 25) {
+    pageNum.innerHTML = nextPage;
+    showsDiv.innerHTML = '';
+    displayShows(shows, nextPage, showsHeader, showsDiv, likesApiURL, body);
+  }
+}
```
Comment (ichala) [inline src/utilities/homePageUtilities.js:99]: - [ ] as part of the requirement, you guys should stick with ES6 as you did with most of your functions, kindly update `function loadNext` to be an arrow function 🙏

File: src/utilities/homePageUtilities.js [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+          <div class='info-btns'>
+            <button class='reservation-btn'>Reservations</button>
+            <button class='comments'>Comments</button>
+          </div>
+          <div class='like-section'>
+          <button class='like-btn'><i class='fa-regular fa-heart' ></i></button>  
+          <span class='likes-number'></span>
+          </div>
+        </div>
+      `;
```
Comment (ichala) [inline src/utilities/homePageUtilities.js:31]: - [ ] as part of the requirement, you guys should stick with ES6 as you did with most of your functions, kindly update `function createShowCard` to be an arrow function 🙏

File: src/utilities/postData.js [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+    headers: {
+      'Content-Type': 'application/json',
+    },
+  })
+    .then((res) => res)
+    .then(() => rerenderList(id, `${url}?item_id=`))
+    .catch((err) => console.log('Network Error: ', err));
+};
+
+export default addNewReservation;
```
Comment (ichala) [inline src/utilities/postData.js:21]: - [ ] for javascript best practices, kindly avoid exporting the same function twice, `export const addNewReservation` & `export default addNewReservation;` , i would suggest removing the `export default addNewReservation;` since your current importing relying on  `export const addNewReservation` , kindly fix it 🙏

File: src/utilities/postData.js [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+      date_start: dateStart,
+      date_end: dateEnd,
+    }),
+    headers: {
+      'Content-Type': 'application/json',
+    },
+  })
+    .then((res) => res)
+    .then(() => rerenderList(id, `${url}?item_id=`))
+    .catch((err) => console.log('Network Error: ', err));
```
Comment (ichala) [inline src/utilities/postData.js:18]: - [ ] for javascript best practices try always to not commit `console.log`  to your repo. It’s ugly, it kills performance and it can make confidential data be visible to anyone using the browser tools to look at your website. kindly remove it and you can see more details in [here](https://github.com/microverseinc/curriculum-html-css/blob/main/articles/javascript_best_practices.md) 🙏

File: src/utilities/renderPopup.js [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+  const dateStart = Popup.querySelector('.startdate-input');
+  const dateEnd = Popup.querySelector('.enddate-input');
+
+  getAllReservations(obj.id, getReservationsURL)
+    .then((res) => res.json())
+    .then((data) => {
+      reservations = data;
+      counter = count(data) || 0;
+      renderReservations(reservations, counter);
+    }).catch((error) => console.log(error));
```
Comment (ichala) [inline src/utilities/renderPopup.js:31]: - [ ] for javascript best practices try always to not commit `console.log`  to your repo. It’s ugly, it kills performance and it can make confidential data be visible to anyone using the browser tools to look at your website. kindly remove it and you can see more details in [here](https://github.com/microverseinc/curriculum-html-css/blob/main/articles/javascript_best_practices.md) 🙏

File: src/tests/reservcounter.test.js [hunk 8]
```diff
@@ -0,0 +1,27 @@
...
+  add(createObj(1, 'kareem', '15-12-2022', '16-12-2022'));
+  expect(count(reservations)).toBe(1);
+});
+
+test('', () => {
+  add(createObj(1, 'kareem', '15-12-2022', '16-12-2022'));
+  add(createObj(2, 'kareem', '15-12-2022', '16-12-2022'));
+  add(createObj(3, 'kareem', '15-12-2022', '16-12-2022'));
+  add(createObj(4, 'kareem', '15-12-2022', '16-12-2022'));
+  expect(count(reservations)).toBe(5);
```
Comment (ichala) [inline src/tests/reservcounter.test.js:26]: - [ ] Good Job testing all of the functions of the counters 🎉 but please note that in `reservcounter.test.js`  you should test and cover all edge cases that can happen, also you tested the same function twice in two different tests block, which is unnecessary, instead I would suggest testing the case of having 0 items in the array as you guys did in the other counters testing, by implementing that you will cover all of the edge cases and you will avoid repeating same tests in this page  `reservcounter.test.js` , also it would be great if you guys group the tests inside a describe method 🙏 kindly fix that

File: src/utilities/commentsPopUpUtilities.js [hunk 9]
```diff
@@ -0,0 +1,65 @@
...
+      <h2>Add comment: </h2>
+      <form class="popup-form" action="#">
+        <input type="text" class="name-input">
+        <textarea name="" id="" class="comment-input" rows="3"></textarea>
+        <button type="submit">Add comment</button>
+      </form>
+    </section>
+    <section class="comments-section">
+      <h2 class="comments-title">Comments: </h2>
+      <ul class="comments-list"></ul>
```
Comment (ichala) [inline src/utilities/commentsPopUpUtilities.js:55]: - [ ] [OPTIONAL] your website looks awesome and I'm really impressed  👏  I would love to suggest a small CSS modification to your comments-list & reservation-list to improve the UI/UX , I would suggest adding a fixed height to your list and add the attribute `overflow-y: scroll;`
 example : 
```CSS
.comments-list {
    display: flex;
    flex-direction: column;
    gap: 10px;
    /* add fixed height or max-height*/
    height: 205px;
    /*add scroll bar to scroll through the list only*/
    overflow-y: scroll;
}

```
**_Result :_** 
OVERFIX

**Review by ichala — CHANGES_REQUESTED** [review]
Hi Team
Good job so far! 
👏👏👏 
There are some issues that you still need to work on to go to the next project but you're almost there!

the-simpsons-homer-simpson

## To Highlight 🎉

- Great Design✔️ 
- No linters Errors ✔️

 ## Required Changes ♻️

- [ ] 🛑  As part of the project requirement, kindly talk about a piece of code that you guys are proud of in the video presentation
- [ ] 🛑  **_Check The comments below_**



## Optional suggestions

- **_Check The comments below_**

**********
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you take them into account as they can make your code better._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Review by ichala — APPROVED** [review]
Hi Team,
# Status: Approved :heavy_check_mark: :heavy_check_mark: :heavy_check_mark:  
 
68747470733a2f2f6d65646961322e67697068792e636f6d2f6d656469612f6c3366516333614356304871707a5763552f67697068792e676966

# To Highlight 💪🏻

- Changes Implemenetd ✔️  
- Good Presentation Video ✔️


 Good job implementing the required changes :+1: Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: 
Congratulations! 🎉 

# Optional Suggestions  :
_ N/A
**********
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._
Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag  @ichala in your question so I can receive the notification.


------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._
